py4DSTEM/process/phase/parameter_optimize.py

```diff
@@ -0,0 +1,449 @@
+        over, the value in the dictionary is used. To optimize a parameter, instead pass an
+        OptimizationParameter object inside the dictionary to specify the initial guess, bounds,
+        and type of parameter, for example:
+            >>> 'param':OptimizationParameter(initial guess, lower bound, upper bound)
+        Calling optimize will then run the optimization simultaneously over all
+        optimization parameters. To obtain the optimized parameters, call get_optimized_arguments
+        to return a set of dictionaries where the OptimizationParameter objects have been replaced
+        with the optimal values. These can then be modified for running a full reconstruction.
+
+        Parameters
+        ----------
+        reconstruction_type: class
+            Type of ptychographic reconstruction to perform
+        init_args: dict
+            Keyword arguments passed to the __init__ method of the reconstruction class
+        preprocess_args: dict
+            Keyword arguments passed to the preprocess method the reconstruction object
+        reconstruction_args: dict
+            Keyword arguments passed to the reconstruct method the the reconstruction object
+        affine_args: dict
+            Keyword arguments passed to AffineTransform. The transform is applied to the initial
+            scan positions.
+        """
+
+        # loop over each argument dictionary and split into static and optimization variables
+        (
+            self._init_static_args,
+            self._init_optimize_args,
+        ) = self._split_static_and_optimization_vars(init_args)
+        (
+            self._affine_static_args,
+            self._affine_optimize_args,
+        ) = self._split_static_and_optimization_vars(affine_args)
+        (
+            self._preprocess_static_args,
+            self._preprocess_optimize_args,
+        ) = self._split_static_and_optimization_vars(preprocess_args)
+        (
+            self._reconstruction_static_args,
+            self._reconstruction_optimize_args,
+        ) = self._split_static_and_optimization_vars(reconstruction_args)
+
+        # Save list of skopt parameter objects and inital guess
+        self._parameter_list = []
+        self._x0 = []
+        for k, v in (
+            self._init_optimize_args
+            | self._affine_optimize_args
+            | self._preprocess_optimize_args
+            | self._reconstruction_optimize_args
+        ).items():
+            self._parameter_list.append(v._get(k))
+            self._x0.append(v._initial_value)
+
+        self._init_args = init_args
+        self._affine_args = affine_args
+        self._preprocess_args = preprocess_args
+        self._reconstruction_args = reconstruction_args
+
+        self._reconstruction_type = reconstruction_type
+
+        self._set_optimizer_defaults()
+
+    def optimize(
+        self,
+        n_calls: int,
+        n_initial_points: int,
+        **skopt_kwargs: dict,
```

Comment:
Error divergence check and log error suggestion (see below):

```suggestion
        check_error_divergence = True,
        return_log_error = True,
        **skopt_kwargs: dict,
```

Reply:
This will be handed in a separate and larger commit that I'll merge with the PR review commits and push later